### PR TITLE
[framework] `NftSafe`

### PR DESCRIPTION
**Pull request by bausano**
## Description 

Adds `nft_safe` and `royalty` modules to `0x2::sui`. 

Alternative to the following MRs:
- https://github.com/MystenLabs/sui/pull/8678
- https://github.com/MystenLabs/sui/pull/9272

This implementation enables simple use cases which were outlined in #8678 while enabling building more complex functionality on top.

This implementation
* uses the same approach for exclusive listing as for normal listing;
* enables listing only with a specific entity (defined by `&UID`);
* enables interface control (e.g. for more advanced options for depositing or transferring NFTs) by wrapping `OwnerCap`;
* enables public listing (anyone can buy for a set price in `SUI` tokens);
* enables middleware approach to transfer rules (string independent logic to approve a transfer such as royalty, allowlists, etc).

We retain the basic building block for royalty enforcement as proposed by @damirka.
The logic is extended, however, by the aforementioned middleware capability.
In the original implementation of the hot potato pattern, different pieces of logic would have been required to know about each other in order to work together on approving a transfer.
With this approach, we can in [nft-protocol] implement `Allowlist` and custom royalty policies by building on top of `TransferCap<T>` objects.
Additionally, we don't have to have our own BPS royalty implementation but can use the one in this MR.

Let's explore the functionality on the following diagrams:

---

simple-pipeline

In this simple scenario, the owner is only interested in collecting a percentage royalty.

1, Some trading contract such as auction gets NFT from the `NftSafe`
2. `NftSafe` creates hot potato object `TransferRequest`
3. `RoyaltyPolicy` performs BPS calculation upon receiving the hot potato and uses its `TransferCap` to sign the transfer request.
4. `NftSafe` consumes the hot potato after verifying that `RoyaltyPolicy` agreed to it via its `TransferCap`,

---

two-step-pipeline

This shows how independent policies can be piped using programmable transactions to sign transfer requests.
This is the sort of logic we want to implement in originbyte's NFT standard.

1, Some trading contract such as auction gets NFT from the `NftSafe`
2. `NftSafe` creates hot potato object `TransferRequest`
3. Policy 1 uses their `TransferCap` to sign the hot potato. This could be e.g. custom royalty computation logic.
4. Policy 2 uses their `TransferCap` (different from the one owned by Policy 1). This could be e.g. allowlist logic or whatever else the collection creator deems sensible to include in the transfer pipeline. 
5. `0x2::nft_safe::NftSafe` consumes the hot potato after verifying that the two policies _both_ agreed to it.

[nft-protocol]: https://github.com/Origin-Byte/nft-protocol

## Test Plan 

Tests are yet to come.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-framework/sources/collectible/royalty.move [hunk 1]
```diff
@@ -0,0 +1,184 @@
+// Copyright (c) Mysten Labs, Inc.
```
Comment (bausano) [inline crates/sui-framework/sources/collectible/royalty.move:1]: From https://github.com/MystenLabs/sui/pull/8678/files#diff-f6a81512479f6e666fd7c7764df0550ada2f82a912906d779d8c0f53ea1d2d7e but with changes

File: crates/sui-adapter-transactional-tests/tests/dynamic_fields/borrow_wrong_type.exp [hunk 2]
```diff
@@ -12,9 +12,9 @@ created: object(107), object(108)
 written: object(106)
```
Comment (bausano) [inline crates/sui-adapter-transactional-tests/tests/dynamic_fields/borrow_wrong_type.exp:13]: @damirka How can I regenerate docs and these changed files?

File: crates/sui-framework/sources/collectible/collectible.move [hunk 3]
```diff
@@ -68,6 +69,8 @@ module sui::collectible {
         otw: OTW, max_supply: Option<u64>, ctx: &mut TxContext
     ): (
         Publisher,
+        nft_safe::TransferPolicy<Collectible<T>>,
```
Comment (bausano) [inline crates/sui-framework/sources/collectible/collectible.move:72]: These can be created at a later point using the publisher. do we want to export them in this method?

**Comment by damirka** [comment]
@bausano closing in favor of #9514. Please, reach out if there's something that I missed and should be there.

**Pull request closed without merging**